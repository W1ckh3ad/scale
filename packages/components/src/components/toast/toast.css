.toast {
  width: 400px;
  display: flex;
  opacity: 0;
  z-index: 1;
  position: fixed;
  background: var(--scl-color-background-standard);
  box-shadow: var(--scl-shadow-level-0);
  box-sizing: border-box;
  border-radius: var(--scl-radius-4);
  flex-direction: column;
  justify-content: space-between;
}
.toast__body {
  padding: var(--scl-spacing-16);
}
.toast__header {
  margin: 0;
  display: flex;
  padding: var(--scl-spacing-16);
  border-bottom: 1px solid var(--scl-color-grey-20);
  justify-content: space-between;
  font-size: var(--scl-font-variant-heading-4-size);
}
.toast__header a {
  cursor: pointer;
}
.toast__header small {
  margin-top: 5px;
  margin-left: 120px;
  font-size: var(--scl-font-variant-smaller-size);
}
.toast__progress {
  left: 0;
  bottom: 0;
  height: var(--scl-spacing-2);
  display: block;
  overflow: hidden;
  position: absolute;
  background: red;
}
