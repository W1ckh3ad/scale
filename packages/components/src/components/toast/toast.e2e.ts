import { newE2EPage } from '@stencil/core/testing';

describe('scale-toast', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-toast>Toast message</scale-toast>');
    const element = await page.find('scale-toast');
    expect(element).toHaveClass('hydrated');
  });
});
