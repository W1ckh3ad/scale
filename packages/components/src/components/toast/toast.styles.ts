import { JssStyle } from 'jss';

export const styles: JssStyle = {
  toast: {
    boxSizing: 'border-box',
    width: '400px',
    position: 'fixed',
    borderRadius: '3px',
    boxShadow: '0 2px 8px rgba(0, 0, 0, 0.26)',
    background: '#fff',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    zIndex: 1,
    opacity: 0,
  },
  toast__body: {
    padding: '1rem',
  },
  toast__header: {
    padding: '1rem',
    borderBottom: '1px solid #ccc',
    fontSize: '1.25rem',
    margin: 0,
    justifyContent: 'space-between',
    display: 'flex',
    '& a': {
      cursor: 'pointer',
    },
    '& small': {
      fontSize: '0.8rem',
      marginLeft: '120px',
      marginTop: '5px',
    },
  },
  toast__progress: {
    background: 'red',
    height: '2px',
    display: 'block',
    position: 'absolute',
    bottom: 0,
    left: 0,
    overflow: 'hidden',
  },
};
