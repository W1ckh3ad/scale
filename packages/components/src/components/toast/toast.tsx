import {
  Component,
  Prop,
  Method,
  h,
  State,
  Element,
  Host,
} from '@stencil/core';
import classNames from 'classnames';
import { formatDistance, subSeconds } from 'date-fns';

@Component({
  tag: 'scale-toast',
  styleUrl: './toast.css',
  shadow: true,
})
export class Toast {
  /** (optional) Toast size */
  @Prop() size?: string = '';
  /** (optional) Toast variant */
  @Prop() variant?: string = '';
  /** (optional) Toast opened */
  @Prop({ reflect: true }) opened?: boolean;
  /** (optional) Toast autohide time */
  @Prop() autoHide?: boolean | number = false;
  /** (optional) Animated toast */
  @Prop() animated?: boolean = true;
  /** (optional) Toast time */
  @Prop() time?: number;
  /** (optional) Toast position at the top */
  @Prop() positionTop?: number = 12;
  /** (optional) Toast position right */
  @Prop() positionRight?: number = 12;
  /** (optional) Toast fade duration */
  @Prop() fadeDuration?: number = 500;
  /** (optional) Injected CSS styles */
  @Prop({ reflect: true }) styles?: string;

  /** (optional) Toast state progress */
  @State() progress: number = 0;
  /** (optional) Toast state height with offset */
  @State() toastHeightWithOffset: number = 0;

  @Element() element: HTMLElement;

  hideToast: boolean = false;
  timerId = null;

  disconnectedCallback() {
    if (this.timerId) {
      clearTimeout(this.timerId);
      this.timerId = null;
      this.opened = false;
      this.progress = 0;
    }
  }

  close = () => {
    clearInterval(this.timerId);
    this.hideToast = true;
    setTimeout(() => {
      this.timerId = null;
      this.opened = false;
      this.progress = 0;
    }, this.fadeDuration);
  };

  getTime = () => {
    const formattedTime =
      this.time &&
      formatDistance(subSeconds(this.time, 3), new Date(), { addSuffix: true });
    return formattedTime;
  };

  setToastTimeout = () => {
    if (this.opened && this.autoHide !== false && !this.timerId) {
      this.timerId = setInterval(() => {
        this.progress += 1 / (this.getAutoHide() / 1000);
        if (this.progress >= 100) {
          this.close();
        }
      }, 10);
    }
  };

  /** Toast method: open() */
  @Method()
  async open() {
    this.opened = true;
    this.hideToast = false;
  }

  render() {
    this.setToastTimeout();
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <style>{this.transitions(this.toastHeightWithOffset)}</style>
        <style>{this.animationStyle(this.toastHeightWithOffset)}</style>

        <div class={this.getCssClassMap()}>
          <div class="toast__header">
            <slot name="header" />

            <small>{this.getTime()}</small>
            <a onClick={this.close}>
              <span aria-hidden="true">&times;</span>
            </a>
          </div>
          {this.autoHide && (
            <div class="toast__progress" style={{ width: `${this.progress}%` }}>
              &nbsp;
            </div>
          )}
          <div class="toast__body">
            <slot />
          </div>
        </div>
      </Host>
    );
  }

  transitions = offset => `
    @keyframes fadeIn {
      from {
        opacity: 0;
        top: -${offset}px;
      }
      to {
        opacity: 1;
        top: ${this.positionTop}px;
      }
    }

    @keyframes fadeOut {
      from {
        opacity: 1;
        top: ${this.positionTop}px;
      }
      to {
        opacity: 0;
        top: -${offset}px;
      }
    }
  `;

  animationStyle = offset => {
    return `
      .toast--show {
        right: ${this.positionRight}px;
        animation: fadeIn ${this.fadeDuration / 1000}s ease-in-out;
        top: ${this.positionTop}px;
        opacity: 1;
      },
      .toast--show {
        right: ${this.positionRight}px;
        animation: fadeOut ${this.fadeDuration / 1000}s ease-in-out;
        top: -${offset}px;
        opacity: 0;
      }
    `;
  };

  getToastHeightWithOffset() {
    const toastHeight = this.element.shadowRoot.querySelector('.toast')
      .scrollHeight;
    this.toastHeightWithOffset = toastHeight + this.positionTop;
  }

  getAutoHide() {
    if (
      typeof this.autoHide === 'number' ||
      typeof this.autoHide === 'string'
    ) {
      return Number(this.autoHide);
    } else {
      return 0;
    }
  }

  getCssClassMap() {
    return classNames(
      'toast',
      this.size && `toast--size-${this.size}`,
      this.variant && `toast--variant-${this.variant}`,
      !!this.opened && 'toast--opened',
      !!!this.hideToast && 'toast--show',
      !!this.hideToast && 'toast--hide'
    );
  }
}
