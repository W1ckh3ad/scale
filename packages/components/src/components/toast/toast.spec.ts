import { newSpecPage } from '@stencil/core/testing';
import { Toast } from './toast';
import { styles } from './toast.styles';
import jss from 'jss';

describe('Toast', () => {
  let element;
  let stylesheet;
  beforeEach(async () => {
    element = new Toast();
    stylesheet = element.stylesheet = jss.createStyleSheet(styles as any);
    jest.useFakeTimers();
    jest.mock('date-fns');
  });

  const components = [Toast];

  const timeStamp = 1540035262000;

  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components,
      html: `<scale-toast>Toast message</scale-toast>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should match snapshot when opened', async () => {
    const page = await newSpecPage({
      components,
      html: `<scale-toast opened=true >Label</scale-toast>`,
    });
    expect(page.root.shadowRoot).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should close the Toast', () => {
    expect(element.opened).toBe(undefined);
    element.close();
    setTimeout(() => {
      expect(element.opened).toBe(false);
    }, 10);
  });

  it('should open the Toast', () => {
    expect(element.opened).toBe(undefined);
    element.open();
    expect(element.opened).toBe(true);
  });

  it('should hide the toast', () => {
    element.autohide = true;
    element.opened = true;
    element.setToastTimeout();

    setTimeout(() => {
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(setTimeout).toHaveBeenLastCalledWith(
        expect.any(Function),
        element.autohideTime
      );
    }, 10);
  });

  it('should not hide the toast', () => {
    element.autohide = false;
    element.opened = false;
    element.setToastTimeout();
    expect(element.autohide).toBe(false);
  });

  it('should cancel the timeout', () => {
    element.myTimeout = 500;
    element.close();

    setTimeout(() => {
      expect(clearTimeout).toHaveBeenCalledTimes(1);
      expect(element.myTimeout).toEqual(undefined);
    }, 10);
  });

  it('should render with default timeformat', () => {
    element.time = timeStamp;
    element.getTime();

    expect(element.time).toBe(timeStamp);
  });

  it('should handle css classes', () => {
    element.customClass = 'custom';
    expect(element.getCssClassMap()).toContain('custom');

    element.variant = 'primary';
    stylesheet.addRule('toast--variant-primary', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['toast--variant-primary']
    );

    element.size = 'small';
    stylesheet.addRule('toast--size-small', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['toast--size-small']
    );

    element.opened = true;
    stylesheet.addRule('toast--opened', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['toast--opened']
    );

    element.hideToast = false;
    stylesheet.addRule('toast--show', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['toast--show']
    );

    element.hideToast = true;
    stylesheet.addRule('toast--hide', {});
    expect(element.getCssClassMap()).toContain(
      stylesheet.classes['toast--hide']
    );
  });
});
