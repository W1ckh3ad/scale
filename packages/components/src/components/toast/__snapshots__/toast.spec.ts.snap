// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should match snapshot 1`] = `
<scale-toast>
  <mock:shadow-root>
    <style>
      .toast {
  width: 400px;
  display: flex;
  opacity: 0;
  z-index: 1;
  position: fixed;
  background: #fff;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.26);
  box-sizing: border-box;
  border-radius: 3px;
  flex-direction: column;
  justify-content: space-between;
}
.toast__body {
  padding: 1rem;
}
.toast__header {
  margin: 0;
  display: flex;
  padding: 1rem;
  border-bottom: 1px solid #ccc;
  justify-content: space-between;
  font-size: 1.25rem;
}
.toast__header a {
  cursor: pointer;
}
.toast__header small {
  margin-top: 5px;
  margin-left: 120px;
  font-size: 0.75rem;
}
.toast__progress {
  left: 0;
  bottom: 0;
  height: 2px;
  display: block;
  overflow: hidden;
  position: absolute;
  background: red;
}
    </style>
    <style>
      @keyframes fadeIn {
      from {
        opacity: 0;
        top: -0px;
      }
      to {
        opacity: 1;
        top: 12px;
      }
    }

    @keyframes fadeOut {
      from {
        opacity: 1;
        top: 12px;
      }
      to {
        opacity: 0;
        top: -0px;
      }
    }
    </style>
    <style></style>
    <div class="toast--show toast">
      <div class="toast__header">
        <slot name="header"></slot>
        header
        <small></small>
        <a>
          <span aria-hidden="true">
            ×
          </span>
        </a>
      </div>
      <div class="toast__body">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  Toast message
</scale-toast>
`;

exports[`Toast should match snapshot when opened 1`] = `
<scale-toast opened="">
  <mock:shadow-root>
    <style>
      .toast {
  width: 400px;
  display: flex;
  opacity: 0;
  z-index: 1;
  position: fixed;
  background: #fff;
  box-shadow: 0 2px 8px rgba(0, 0, 0, 0.26);
  box-sizing: border-box;
  border-radius: 3px;
  flex-direction: column;
  justify-content: space-between;
}
.toast__body {
  padding: 1rem;
}
.toast__header {
  margin: 0;
  display: flex;
  padding: 1rem;
  border-bottom: 1px solid #ccc;
  justify-content: space-between;
  font-size: 1.25rem;
}
.toast__header a {
  cursor: pointer;
}
.toast__header small {
  margin-top: 5px;
  margin-left: 120px;
  font-size: 0.75rem;
}
.toast__progress {
  left: 0;
  bottom: 0;
  height: 2px;
  display: block;
  overflow: hidden;
  position: absolute;
  background: red;
}
.toast--show {
  top: 12px;
  right: 12px;
  opacity: 1;
  animation: fadeIn 0.5s ease-in-out;
}
.toast--hide {
  top: -0px;
  right: 12px;
  opacity: 0;
  animation: fadeOut 0.5s ease-in-out;
}
    </style>
    <style>
      @keyframes fadeIn {
      from {
        opacity: 0;
        top: -0px;
      }
      to {
        opacity: 1;
        top: 12px;
      }
    }

    @keyframes fadeOut {
      from {
        opacity: 1;
        top: 12px;
      }
      to {
        opacity: 0;
        top: -0px;
      }
    }
    </style>
    <style></style>
    <div class="toast--show toast">
      <div class="toast__header">
        <slot name="header"></slot>
        header
        <small></small>
        <a>
          <span aria-hidden="true">
            ×
          </span>
        </a>
      </div>
      <div class="toast__body">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-toast>
`;
